File: frontend/src/components/project/statics.js
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getProjectStatics } from '../../actions/projects';
import StaticsItem from './static-item';

class ProjectStatics extends Component {
    constructor(props) {
        super(props)
        this.state = {
            username: this.props.user
        }

        /* receive project statics details from server */
        this.props.getProjectStatics(this.props.project_data.id)
    }

    render() {
        if (this.props.statics == null)
            return ''
        return (
            <div id="project-statics" className="section-in-main">
                <div className="header-section-v1 header-v1-yellow">
                    <h1 id="projects-title">
                        Project Statics
                    </h1>
                    <h2 id="projects-intro">
                        this sections display project statics
                    </h2>
                </div>
                <div className="container">
                    <div id="files-statics">
                        <StaticsItem color="cyan" category="Number of uploaded files" value={ this.props.statics.files.files_quantity }  />
                        <StaticsItem color="cyan" category="Total size of uploaded files" value={ this.props.statics.files.total_size }  />
                        <StaticsItem color="cyan" category="Most File Upload Times" value={ '' }  />
                        <StaticsItem color="cyan" type="most-favorite" id="insert_by" category="Member" value={ this.props.statics.files.most_uploader }  />
                        <StaticsItem color="cyan" type="most-favorite" id="insertion_date" category="Date" value={ this.props.statics.files.most_file_uploads }  />
                    </div>
                    <p/>
                    <div id="model-statics">
                        <StaticsItem color="purple" category="Parameters quantity" value={ this.props.statics.model.parameters }  />
                        <StaticsItem color="purple" category="Layers quantity" value={ this.props.statics.model.layers }  />
                        <StaticsItem color="purple" category="Modules Usage" value={ '' }  />
                        <StaticsItem color="purple" category="Dropout" value={  this.props.statics.model.dropout + " modules" }  />
                        <StaticsItem color="purple" category="BatchNorm" value={  this.props.statics.model.batchnorm + " modules" }  />
                        <StaticsItem color="purple" category="Convolution" value={  this.props.statics.model.convolution + " modules" }  />
                        <StaticsItem color="purple" category="Activaions" value={  this.props.statics.model.activations.quantity + " modules" }  />
                        <StaticsItem color="purple" category="Max Used Activation" value={  this.props.statics.model.activations.max_used + " ( rate " + this.props.statics.model.activations.value + " ) " }  />

                    </div>
                    <p/>
                    <div id="team-statics">
                        <StaticsItem color="green" category="Team members quantity" value={ this.props.statics.team.team_size }  />
                        <StaticsItem color="green" category="Project managers quantity" value={ this.props.statics.team.managers }  />
                    </div>
                    <p/>
                    <div id="runs-statics">
                        <StaticsItem color="red" category="Total number of runs" value={ this.props.statics.runs.qunatity }  />
                        <StaticsItem color="red" category="Total number of epochs" value={ this.props.statics.runs.total_epoches }  />
                        <StaticsItem color="red" category="Runs in progress" value={ this.props.statics.runs.in_run }  />
                        <StaticsItem color="red" category="Average accuracy rate" value={ this.props.statics.runs.accuracy_average * 100 }  />
                        <StaticsItem color="red" category="Maximum accuracy rate" value={ this.props.statics.runs.accuracy_max * 100 }  />
                        <StaticsItem color="red" category="Minimum accuracy rate" value={ this.props.statics.runs.accuracy_min * 100 }  />
                        <StaticsItem color="red" category="Average loss value" value={ this.props.statics.runs.loss_average }  />
                        <StaticsItem color="red" category="Maximum loss value" value={ this.props.statics.runs.loss_max }  />
                        <StaticsItem color="red" category="Minimum loss value" value={ this.props.statics.runs.loss_min }  />
                        <StaticsItem color="red" category="Underfitting Runs" value={ this.props.statics.runs.underfitting.quantity }  />
                        <StaticsItem color="red" category="Underfitting Rate" value={ this.props.statics.runs.underfitting.rate }  />
                        <StaticsItem color="red" category="Overfitting Runs" value={ this.props.statics.runs.overfitting.quantity }  />
                        <StaticsItem color="red" category="Overfitting Rate" value={ this.props.statics.runs.overfitting.rate }  />
                        <StaticsItem color="red" category="Favorite Hyper Parameters" value={ '' }  />
                        <StaticsItem color="red" type="most-favorite" id="epochs" category="Epochs" value={ this.props.statics.runs.favorite_params.epochs }  />
                        <StaticsItem color="red" type="most-favorite" id="batch_size" category="Batch Size" value={ this.props.statics.runs.favorite_params.batch_size }  />
                        <StaticsItem color="red" type="most-favorite" id="learning_rate"  category="Learning Rate" value={ this.props.statics.runs.favorite_params.learning_rate }  />
                        <StaticsItem color="red" type="most-favorite" id="weight_decay" category="Weight Decay" value={ this.props.statics.runs.favorite_params.weight_decay }  />
                        <StaticsItem color="red" type="most-favorite" id="optimizer" category="Optimizer" value={ this.props.statics.runs.favorite_params.optimizer }  />
                        <StaticsItem color="red" type="most-favorite" id="loss_type" category="Loss Type" value={ this.props.statics.runs.favorite_params.loss }  />
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        project_data: state.projectReducer.project_selected,
        statics: state.projectReducer.project_selected.statics
    }
}

const mapDispatchToProps = dispatch => {
    return {
        getProjectStatics: (project_id) => {
            dispatch(getProjectStatics(project_id))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ProjectStatics);

File: backend/actions/statics.py
from backend.analyze.recommendations import Recommendations
from backend.actions.general import float_precision
from backend.actions.runs import favorite_parameter
from backend.submodels.model import ProjectRuns
from backend.submodels.project import ProjectFiles, ProjectTeam
from backend.actions.project import get_total_size, most_query
from django.db.models import Avg, Max, Min, Sum, Count

class ProjectStatics():
    def __init__(self, project):
        self.project = project
        self.recommendations_object = Recommendations(project.id)
        self.analysis = self.recommendations_object.analysis()

    def get_statics(self):
        return {
            'model': self.model_statics(),
            'runs': self.runs_statics(),
            'team': self.team_statics(),
            'files': self.files_statics(),
        }

    def model_statics(self):
        return {
            'parameters': self.analysis['model']['structure']['parameters']['quantity'],
            'layers': self.analysis['model']['structure']['layers_quantity']['quantity'],
            'batchnorm': self.analysis['model']['modules']['batch_normalization']['exist'],
            'dropout': self.analysis['model']['modules']['dropout']['exist'],
            'convolution': self.analysis['model']['modules']['convolution']['quantity']['value'],
            'activations': {
                'max_used': self.analysis['model']['modules']['activations']['max_used']['activation'],
                'value': self.analysis['model']['modules']['activations']['max_used']['value'],
                'quantity': self.analysis['model']['modules']['activations']['usage']['activations_quantity']
            },
        }

    def runs_statics(self):
        results = ProjectRuns.objects.filter(project=self.project).filter(progress=100)
        accuracy_decimal_digits = 6
        loss_decimal_digits = 4
        statics = {
            # add time sum, time average, time max and min
            'qunatity': results.count(),
            'in_run': ProjectRuns.objects.filter(project=self.project).filter(progress__lt=100).count(),
            'total_epoches': results.aggregate(Sum('epochs'))['epochs__sum'],
            'accuracy_average': float_precision(results.aggregate(Avg('accuracy'))['accuracy__avg'], accuracy_decimal_digits),
            'loss_average': float_precision(results.aggregate(Avg('loss'))['loss__avg'], loss_decimal_digits),
            'accuracy_max': float_precision(results.aggregate(Max('accuracy'))['accuracy__max'], accuracy_decimal_digits),
            'accuracy_min': float_precision(results.aggregate(Min('accuracy'))['accuracy__min'], accuracy_decimal_digits),
            'loss_min': float_precision(results.aggregate(Min('loss'))['loss__min'], loss_decimal_digits),
            'loss_max': float_precision(results.aggregate(Max('loss'))['loss__max'], loss_decimal_digits),
            'favorite_params': {
                'batch_size': favorite_parameter(results, 'batch_size'),
                'loss': favorite_parameter(results, 'loss_type'),
                'optimizer': favorite_parameter(results, 'optimizer'),
                'weight_decay': favorite_parameter(results, 'weight_decay'),
                'learning_rate': favorite_parameter(results, 'learning_rate'),
                'epochs': favorite_parameter(results, 'epochs'),
            },
            'most_runner': favorite_parameter(results, 'user'),
            'date_most_runs': favorite_parameter(results, 'date'),
            'overfitting': self.analysis['runs']['runs_status']['overfitting'],
            'underfitting': self.analysis['runs']['runs_status']['underfitting'],
        }
        return statics

    def files_statics(self):
        files = ProjectFiles.objects.filter(project=self.project.id)
        if files.count() > 0:
            most_uploader = most_query(files, 'insert_by')
            most_file_uploads = most_query(files, 'insertion_date')
        else:
            most_uploader = { 'insert_by': 'None', 'count': 0 }
            most_file_uploads = { 'insertion_date': 'None', 'count': 0 }
        statics = {
            'files_quantity' : files.count(),
            'total_size': get_total_size(files),
            'most_uploader': most_uploader,
            'most_file_uploads': most_file_uploads,
        }
        return statics
   
    def team_statics(self):
        team = ProjectTeam.objects.filter(project=self.project)
        statics = {
            "team_size": team.count(),
            "managers": team.filter(presmissions=5).count()
        }
        return statics